src/graph/sampler.cc

```diff
@@ -988,415 +988,769 @@ std::vector<dgl_id_t> Global2Local(const std::vector<size_t> &ids,
-                            int neg_sample_size, bool exclude_positive,
-                            bool check_false_neg) {
-  int64_t num_tot_nodes = gptr->NumVertices();
-  if (neg_sample_size > num_tot_nodes)
-    neg_sample_size = num_tot_nodes;
-  bool is_multigraph = gptr->IsMultigraph();
-  std::vector<IdArray> adj = pos_subg.graph->GetAdj(false, "coo");
-  IdArray coo = adj[0];
-  int64_t num_pos_edges = coo->shape[0] / 2;
-  int64_t num_neg_edges = num_pos_edges * neg_sample_size;
-  IdArray neg_dst = IdArray::Empty({num_neg_edges}, coo->dtype, coo->ctx);
-  IdArray neg_src = IdArray::Empty({num_neg_edges}, coo->dtype, coo->ctx);
-  IdArray neg_eid = IdArray::Empty({num_neg_edges}, coo->dtype, coo->ctx);
-  IdArray induced_neg_eid = IdArray::Empty({num_neg_edges}, coo->dtype, coo->ctx);
-
-  // These are vids in the positive subgraph.
-  const dgl_id_t *dst_data = static_cast<const dgl_id_t *>(coo->data);
-  const dgl_id_t *src_data = static_cast<const dgl_id_t *>(coo->data) + num_pos_edges;
-  const dgl_id_t *induced_vid_data = static_cast<const dgl_id_t *>(pos_subg.induced_vertices->data);
-  const dgl_id_t *induced_eid_data = static_cast<const dgl_id_t *>(pos_subg.induced_edges->data);
-  size_t num_pos_nodes = pos_subg.graph->NumVertices();
-  std::vector<size_t> pos_nodes(induced_vid_data, induced_vid_data + num_pos_nodes);
+inline SubgraphRef ConvertRef(const Subgraph &subg) {
+    return SubgraphRef(std::shared_ptr<Subgraph>(new Subgraph(subg)));
+}
 
-  dgl_id_t *neg_dst_data = static_cast<dgl_id_t *>(neg_dst->data);
-  dgl_id_t *neg_src_data = static_cast<dgl_id_t *>(neg_src->data);
-  dgl_id_t *neg_eid_data = static_cast<dgl_id_t *>(neg_eid->data);
-  dgl_id_t *induced_neg_eid_data = static_cast<dgl_id_t *>(induced_neg_eid->data);
-
-  const dgl_id_t *unchanged;
-  dgl_id_t *neg_unchanged;
-  dgl_id_t *neg_changed;
-  if (IsNegativeHeadMode(neg_mode)) {
-    unchanged = dst_data;
-    neg_unchanged = neg_dst_data;
-    neg_changed = neg_src_data;
-  } else {
-    unchanged = src_data;
-    neg_unchanged = neg_src_data;
-    neg_changed = neg_dst_data;
+inline SubgraphRef ConvertRef(const NegSubgraph &subg) {
+    return SubgraphRef(std::shared_ptr<Subgraph>(new NegSubgraph(subg)));
+}
+
+}  // namespace
+
+DGL_REGISTER_GLOBAL("sampling._CAPI_GetNegEdgeExistence")
+.set_body([] (DGLArgs args, DGLRetValue* rv) {
+  SubgraphRef g = args[0];
+  auto gptr = std::dynamic_pointer_cast<NegSubgraph>(g.sptr());
+  *rv = gptr->exist;
+});
+
+DGL_REGISTER_GLOBAL("sampling._CAPI_GetEdgeSubgraphHead")
+.set_body([] (DGLArgs args, DGLRetValue* rv) {
+  SubgraphRef g = args[0];
+  auto gptr = std::dynamic_pointer_cast<NegSubgraph>(g.sptr());
+  *rv = gptr->head_nid;
+});
+
+DGL_REGISTER_GLOBAL("sampling._CAPI_GetEdgeSubgraphTail")
+.set_body([] (DGLArgs args, DGLRetValue* rv) {
+  SubgraphRef g = args[0];
+  auto gptr = std::dynamic_pointer_cast<NegSubgraph>(g.sptr());
+  *rv = gptr->tail_nid;
+});
+
+class EdgeSamplerObject: public Object {
```

Comment:
what is `Object`?

Reply:
See include/dgl/runtime/object.h This is used for ffi object explored to python